### PR TITLE
les: implement request distributor, fix blocking issues

### PR DESCRIPTION
**Pull request by zsfelfoldi**
This PR implements a new, cleaner and more efficient LES request distribution mechanism that also avoids starvation issues by prioritizing requests that have been created earlier. Also, it fixes a minor issue in light.TxPool.

### CODE REVIEW

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 66ee7591f02a6b0c751d7672a6d50f79306b48b8
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by mchusovlianov** [comment]
I've tested a bit this PR ( I just run ./build/bin/geth --light --testnet console and i try to dump eth object by typing eth ENTER):
image

**Comment by zsfelfoldi** [comment]
thanks for the bug report, I will fix it soon.

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 66ee7591f02a6b0c751d7672a6d50f79306b48b8
  * Commits must be prefixed with the package(s) they modify

 * Commit: c4217554e5757a2c3c3fc56d49960241953a0dff
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by zsfelfoldi** [comment]
fixed

**Comment by mchusovlianov** [comment]
It looks now it has deadlock:
image

image

**Comment by zsfelfoldi** [comment]
Not a deadlock, the server was disconnected. Did it happen because of your 'eth' command?

**Comment by farazdagi** [comment]
- I can confirm the same, if you type 'eth, eth, .. , eth N' at some (not very high) N, everything blocks, and nothing happens, smells lots like deadblock

**Comment by zsfelfoldi** [comment]
Does it say "Rolled back 2048 headers..." like on the picture above? That means server disconnection during syncing. Which is bad too if an 'eth' command causes that, but it is a different issue.

**Comment by farazdagi** [comment]
- yep, it always (at least in my anecdotical evidence) happens in sequence: "eth" strings start to pile, and then "rolled back" message appears, and while I can input anything, no response from server ever comes
- Additionally, when I had another console attached in a separate window, it also just locks, when "eth DOS" happens in the first window.

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 314c919b98f1ebee6212265806116f92ee5006be
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 314c919b98f1ebee6212265806116f92ee5006be
  * Commits must be prefixed with the package(s) they modify

 * Commit: 297ccfb780bd916ccc91f7911bdb50e5b19448ff
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by zsfelfoldi** [comment]
Please try it again... there is still some weird issue when resuming a cancelled syncing (after header rollback), I will debug that later, but the really nasty bug that caused blocking has been found.

**Comment by farazdagi** [comment]
Thank you! The issue seems to be resolved - at least I couldn't overwhelm console with "eth" calls. Rollbacks happen - yet sync resumes and full sync happens, eventually.

So, all in all - this nasty blocking bug has been resolved.

**Comment by farazdagi** [comment]
I've done more tests, unfortunately got the following panic:
image

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 314c919b98f1ebee6212265806116f92ee5006be
  * Commits must be prefixed with the package(s) they modify

 * Commit: 297ccfb780bd916ccc91f7911bdb50e5b19448ff
  * Commits must be prefixed with the package(s) they modify

 * Commit: fe38042772b96c5132d272d4534a75f179e4e6f7
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by zsfelfoldi** [comment]
Thanks for catching it, fixed. Now I am carefully checking all my locks, callbacks and goroutines :)

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 314c919b98f1ebee6212265806116f92ee5006be
  * Commits must be prefixed with the package(s) they modify

 * Commit: 297ccfb780bd916ccc91f7911bdb50e5b19448ff
  * Commits must be prefixed with the package(s) they modify

 * Commit: fe38042772b96c5132d272d4534a75f179e4e6f7
  * Commits must be prefixed with the package(s) they modify

 * Commit: 5afd26a97148d6d460f446c4ad2585d57dc0857e
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 314c919b98f1ebee6212265806116f92ee5006be
  * Commits must be prefixed with the package(s) they modify

 * Commit: 297ccfb780bd916ccc91f7911bdb50e5b19448ff
  * Commits must be prefixed with the package(s) they modify

 * Commit: fe38042772b96c5132d272d4534a75f179e4e6f7
  * Commits must be prefixed with the package(s) they modify

 * Commit: 5afd26a97148d6d460f446c4ad2585d57dc0857e
  * Commits must be prefixed with the package(s) they modify

 * Commit: c4f82a1ea903288f8356f4ca32f24f82f9eaa00c
  * Commits must be prefixed with the package(s) they modify

 * Commit: b4be9ed6160daab549b8eafbbe75ec26a6c25858
  * Commits must be prefixed with the package(s) they modify

 * Commit: aee030fa7a6485997ddd166544478ffdecb911cc
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by zsfelfoldi** [comment]
also fixed a bug that caused flow control errors and random peer disconnections.

**Comment by farazdagi** [comment]
While overall blocking is gone indeed, there's one minor issue with a delay, rather than a block.

I've done the following experiment:
- wiped out chaindata (so that the test starts from trusted CHT, and needs to sync all the way from it)
- send couple of transactions
  - sometimes those transactions validate just fine, even if actual nonce is higher (in the last block), it might so happen that when tx is sent, chain state doesn't know about those higher nonces
  - curiously enough, the first transaction proceeds fine (of course it never gets included into blockchain, as peers deny it)
  - however, the second take *much* longer to complete (like minutes, instead of seconds)

I've tracked the problem to https://github.com/ethereum/go-ethereum/blob/master/light/txpool.go#L303:

```golang
func (pool *TxPool) eventLoop() {
	for ev := range pool.events.Chan() {
		switch ev.Data.(type) {
		case core.ChainHeadEvent:
			pool.mu.Lock()
			ctx, _ := context.WithTimeout(context.Background(), blockCheckTimeout)
			head := pool.chain.CurrentHeader()
			txc, _ := pool.setNewHead(ctx, head)
			m, r := txc.getLists()
			pool.relay.NewHead(pool.head, m, r)
			pool.homestead = pool.config.IsHomestead(head.Number)
			pool.signer = types.MakeSigner(pool.config, head.Number)
			pool.mu.Unlock()
                        // <<<<<< delay is solved if I add time.Sleep(0) or runtime.Gosched()
		}
	}
}
```

So, `eventLoop` acquires lock and preemption seems to get greatly delayed i.e. another routine was starving on `txpool.Add()` lock acquiring - hence the delay.

Once I added `runtim.Gosched()` to allow for manual preemption and context switching, no more delays occurred.

**Comment by farazdagi** [comment]
**Further update:**
- we had to downgrade to 1.5.5-unstable (to commit before https://github.com/ethereum/go-ethereum/pull/3413/commits/c57c54ce96628aeb6345776310123a80593f0143), as LES sync is not as fast on mobile devices as it is on desktop - and blocknig/delays during sync seriously impedes UX
- before mentioned commit, it seems that RPC requests were serviced using local database, right now it seems that they are all network/peer requests + contend for CPU time with syncing routines, hence delays
- for example, in https://github.com/ethereum/go-ethereum/blob/master/les/odr.go#L233 doc comment claims that we first try local storage, and only then rely on network. But the code doesn't follow this - we are sending requests to peers, always? Given that there are multiples of trie retrieval requests per RPC call (even `eth_getBalance()` will result in multiple calls), we need responsive peers to get results in a timely manner
- another problem is not servicing RPC calls before sync starts - which is not user-friendly as well. You see, we can have lots of data in local db, and it is not clear why call to `eth_getBalance()` needs to wait up until we locate peers and start syncing? Previously (again the same commit), it wasn't like that. Instead, you can get response immediately (your result might be some outdated info, of course - as sync hasn't yet been started and concluded)
- to sum up, while request distributor is a nice addition, it **degrades the responsiveness** of LES server up until sync is (almost) done. Previously, you could have received slightly outdated results, but receive them pretty quickly. Now, a single RPC call may incur up to some 1-2 minutes of delay, on average. And this delays simply kill all the UX on mobile devices, where LES sync process has always be noticeably slower than on desktop

**Comment by zsfelfoldi** [comment]
@farazdagi thanks for reporting these issues, I've switched the status of the PR back to "in progress" until I address them.

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: f93567b42de22c77f7a08afe955b21ea0de067eb
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by zsfelfoldi** [comment]
@farazdagi I have investigated the mentioned issues and found the following:
- if you send a transaction before syncing up, the light transaction pool kills performance by downloading each block until it finds the sent transaction. It also wastes some CPU performance while syncing, even if no tx has been sent. This problem will be fixed by a subsequent PR.
- serving data from the local db before syncing is started DOES work if everything is there, or at least I could not reproduce a scenario where it does not. The comments above the Retrieve funcion were outdated, db checking is done in the access functions found in light/odr_util.go.
- the commit https://github.com/ethereum/go-ethereum/pull/3413/commits/c57c54ce96628aeb6345776310123a80593f0143 you mentioned only affected the server mode and was rolled back since then, so it can not be related to these problems.

Hopefully the remaining performance and responsiveness issues will be fixed by the new (and actually simpler) light tx pool which will (again, hopefully) be included in the 1.6 release.

**Comment by farazdagi** [comment]
@zsfelfoldi thank you for providing more details. I've mentioned this on Gitter, copying it here:

> in my tests https://github.com/ethereum/go-ethereum/pull/3660 still doesn't fix the blocking. I did the following: 1. reset chaindata 2. start node 3. in another window attach to that node 4. try to `eth.getBalance()` and `eth.estimateGas()` 5. blocking/lags occurs.

So, unfortunately, for us it this remaining performance and responsiveness issues were show stoppers (as on mobile LES is already understandably slower than that on desktop). What we did is simply excluded any updates that happened from https://github.com/ethereum/go-ethereum/pull/3413/commits/c57c54ce96628aeb6345776310123a80593f0143 I literally cherry-picked every other commits, but ones related to LES, and we have non blocking and almost up to date Geth (everything is up to date, except for LES).

I am not sure how https://github.com/ethereum/go-ethereum/pull/3413/commits/c57c54ce96628aeb6345776310123a80593f0143 is unrelated to the issue - I see no blocking before that commit, I see no blocking if I update everything else bug LES. It is possible that commit has been reverted, and then something else **re-introduced** the issue, I am not sure.

At the moment we have our hackish way to keep updated Geth and outdated LES 😞  - once I have more time, I will look further into issue, as the original issue I've reported (https://github.com/ethereum/go-ethereum/issues/3639) is definitely not fixed by this PR, so hopefully it gets addressed down the road.

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 75d40abff20173e60e8b4196b8993ff6aa948cf9
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 75d40abff20173e60e8b4196b8993ff6aa948cf9
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

File: les/distributor.go [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+// distPeer is an LES server peer interface for the request distributor.
+// waitBefore returns either the necessary waiting time before sending a request
+// with the given upper estimated cost or the estimated remaining relative buffer
+// value after sending such a request (in which case the request can be sent
+// immediately). At least one of these values is always zero.
+type distPeer interface {
+	waitBefore(uint64) (time.Duration, float64)
+	canSendOrdered() bool
+	orderedSend(f func())
+}
```
Comment (fjl) [inline les/distributor.go]: As discussed, please remove the need for 'ordered sends' by updating flow control counters when the request is actually sent.

File: les/distributor.go [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+			for req != nil {
+				close(req.sentChn)
+				req = req.queueNext
+			}
+			d.lock.Unlock()
+			return
+		case retry := <-d.loopChn:
+			// the main loop distinguishes two kinds of signals:
+			// - "retry" is sent to re-check whether previously unsendable requests can now be sent to one of the peers
+			// - "next" is sent when there are more requests to be processed/sent
```
Comment (fjl) [inline les/distributor.go]: Please remove the retry mechanism.

File: les/distributor.go [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+			r.queueNext = before
+			r.queuePrev = before.queuePrev
+			r.queueNext.queuePrev = r
+			if r.queuePrev == nil {
+				d.queueFirst = r
+			} else {
+				r.queuePrev.queueNext = r
+			}
+		}
+	}
```
Comment (fjl) [inline les/distributor.go]: You might be able to remove this part if you use container/list.

File: les/distributor.go [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+	return r
+}
+
+// newDistReq creates a new request instance that can be queued for distribution
+func newDistReq(getCost func(distPeer) uint64, canSend func(distPeer) bool, request func(distPeer) func()) *distReq {
+	return &distReq{
+		getCost: getCost,
+		canSend: canSend,
+		request: request,
+	}
```
Comment (fjl) [inline les/distributor.go]: Please change callers to use `&distReq{getCost: func(...) uint64 { ... }, ...}` instead of calling newDistReq. Doing this will add "labels" to all the func literals.

File: les/distributor.go [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+		req = next
+	}
+
+	if sel != nil {
+		c := sel.choose().(selectPeerItem)
+		return c.peer, c.req, 0
+	}
+	if bestReq == nil {
+	} else {
+	}
```
Comment (fjl) [inline les/distributor.go]: You don't need this empty condition anymore ;)

File: les/distributor.go [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+			for {
+				peer, req, wait := d.nextRequest()
+				if req != nil && wait == 0 {
+					chn := req.sentChn
+					d.remove(req)
+					send := req.request(peer)
+					if send != nil {
+						peer.orderedSend(send)
+					}
+					chn <- peer
```
Comment (fjl) [inline les/distributor.go]:
```go
req.sentChn <- peer
close(req.sentChn)
```

File: les/distributor.go [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+			fmt.Println("running", d.loopNextSent, "wakeup", d.loopRetrySent)
+			r := d.queueFirst
+			for r != nil {
+				var p uint64
+				if r.queuePrev != nil {
+					p = r.queuePrev.reqOrder
+				}
+				fmt.Println(" order", r.reqOrder, "prev", p)
+				r = r.queueNext
+			}
```
Comment (fjl) [inline les/distributor.go]: Please remove this select clause.

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 4f3a9eea6aff8d15aa7ce328af343e32c6bc653e
  * Commits must be prefixed with the package(s) they modify

 * Commit: 73f68d563c1fa1d0e01cedcebbe91726b2ace6de
  * Commits must be prefixed with the package(s) they modify

 * Commit: 04067f3fffb9a0ef4bf36604ec197be79cb8c337
  * Commits must be prefixed with the package(s) they modify

 * Commit: 9a82a89c338754e1b5bd0d3542ca9499ba473d7b
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

File: les/distributor.go (hunk 1 above)
Comment (zsfelfoldi) [inline les/distributor.go]: I realized that I'd really want to keep the ordered send because the order actually affects performance, the optimal thing to do is to keep the queue priority and always send the oldest requests first. I know this is a minor nuance but with the recent additions to the protocol there are cases where the order of reply messages counts too, so I believe in the long run this is the cleanest solution. So I made the ordered send a general thing and put it into the common package.

File: les/distributor.go (hunk 6 above)
Comment (zsfelfoldi) [inline les/distributor.go]: No, I realized that remove sets sentChn to nil so there was actually a reason for this.

**Comment by zsfelfoldi** [comment]
@fjl I did the requested changes (except two of them, see my comments).

File: common/execqueue.go [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+		return false
+	}
+	q.chn <- f
+	return true
+}
+
+// Stop stops the exec queue.
+func (q *ExecQueue) Stop() {
+	atomic.StoreInt32(&q.stop, 1)
+}
```
Comment (fjl) [inline common/execqueue.go]: Please add this as an unexported utility in package les instead. We don't want to add new features to package common anymore.

Comment (zsfelfoldi) [inline common/execqueue.go]: done

File: light/lightchain.go [hunk 9]
```diff
@@ -369,9 +370,17 @@ func (self *LightChain) postChainEvents(events []interface{}) {
...
+	if i, err := self.hc.ValidateHeaderChain(chain, checkFreq); err != nil {
+		return i, err
+	}
+
 	// Make sure only one thread manipulates the chain at once
 	self.chainmu.Lock()
-	defer self.chainmu.Unlock()
+	defer func() {
+		self.chainmu.Unlock()
+		time.Sleep(time.Millisecond * 10) // ugly hack; do not hog chain lock in case syncing is CPU-limited by validation
```
Comment (fjl) [inline light/lightchain.go]: This is indeed ugly and we should remove it.

File: les/distributor.go [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+						break loop
+					}
+					d.loopNextSent = true // a "next" signal has been sent, do not send another one until this one has been received
+					if wait > distMaxWait {
+						// waiting times may be reduced by incoming request replies, if it is too long, recalculate it periodically
+						wait = distMaxWait
+					}
+					go func() {
+						time.Sleep(wait)
+						d.loopChn <- struct{}{}
```
Comment (fjl) [inline les/distributor.go]: I have a better idea for this, will submit after the merge.
